src/Utilities/cachingoptimizer.jl

```diff
@@ -10,52 +10,114 @@
+  in the incorrect state results in an error.
 * `AUTOMATIC`: The `CachingOptimizer` changes its state when necessary. For
   example, `optimize!` will automatically call `attach_optimizer` (an
   optimizer must have been previously set). Attempting to add a constraint or
   perform a modification not supported by the optimizer results in a drop to
   `EMPTY_OPTIMIZER` mode.
 """
-mutable struct CachingOptimizer{OptimizerType,ModelType<:MOI.ModelLike} <:
-               MOI.AbstractOptimizer
+mutable struct CachingOptimizer{
+    OptimizerType,
+    ModelType<:MOI.ModelLike,
+} <: MOI.AbstractOptimizer
     optimizer::Union{Nothing,OptimizerType}
     model_cache::ModelType
     state::CachingOptimizerState
     mode::CachingOptimizerMode
     model_to_optimizer_map::IndexMap
     optimizer_to_model_map::IndexMap
-    # CachingOptimizer externally uses the same variable and constraint indices
-    # as the model_cache. model_to_optimizer_map maps from the model_cache indices to the
-    # optimizer indices.
+    bridge_constraints::Bool
 end
 
+"""
+    CachingOptimizer(
+        model_cache::MOI.ModelLike,
+        optimizer::Union{Nothing,MOI.AbstractOptimizer} = nothing;
+        mode::CachingOptimizerMode = AUTOMATIC,
+        state::CachingOptimizerState =
+            optimizer === nothing ? NO_OPTIMIZER : ATTACHED_OPTIMIZER,
+        bridge_constraints::Bool = false,
+    )
+
+Creates an `CachingOptimizer` using `model_cache` and `optimizer`.
+
+## Notes
+
+ * If `bridge_constraints == true`, when the caching optimizer encounters a
+   constriant that is not supported by `optimizer`, it automatically adds a
+   bridging layer to `optimizer`.
+ * If `bridge_constraints == true`, the state is forced to `EMPTY_OPTIMIZER`.
+ * If an `optimizer` is passed, the returned CachingOptimizer does not support
+   the function `reset_optimizer(model, new_optimizer)` if the type of
+   `new_optimizer` is different from the type of `optimizer`.
+
+## Examples
+
+```julia
+model = MOI.Utilities.CachingOptimizer(
+    MOI.Utilities.Model{Float64}(),
+    GLPK.Optimizer(),
+)
+```
+
+```julia
+model = MOI.Utilities.CachingOptimizer(
+    MOI.Utilities.Model{Float64}(),
+    bridge_constraints = true,
+)
+MOI.Utilities.reset_optimizer(model, GLPK.Optimizer())
+```
+"""
 function CachingOptimizer(
     model_cache::MOI.ModelLike,
-    mode::CachingOptimizerMode,
+    optimizer::Union{Nothing,MOI.AbstractOptimizer} = nothing;
+    mode::CachingOptimizerMode = AUTOMATIC,
+    state::CachingOptimizerState =
+        optimizer === nothing ? NO_OPTIMIZER : ATTACHED_OPTIMIZER,
+    bridge_constraints::Bool = false,
 )
-    return CachingOptimizer{MOI.AbstractOptimizer,typeof(model_cache)}(
-        nothing,
+    T = typeof(optimizer)
+    if optimizer !== nothing
+        @assert MOI.is_empty(model_cache)
+        @assert MOI.is_empty(optimizer)
+        if bridge_constraints
+            state = EMPTY_OPTIMIZER
+            T = MOI.AbstractOptimizer
```

Comment:
The type is not concrete so it will be inefficient

Reply:
We could use a union of the type with and without bridges

Reply:
It's not sufficient to do this because the bridges may need a caching layer below. The abstract type has worked fine for JuMP, and having a complicated union may not actually help things.

Reply:
I don't agree, we get an allocation everytime we need to access the field, we have all these `_moi` function barriers in JuMP to avoid this. If we have the same issue in the `CachingOptimizer`, then it doesn't make sense.

Reply:
This is an opt-in feature for users of CachingOptimizer. JuMP will opt-in with no change to performance because it already has the abstract type. If others opt-in, they should check performance and/or implement function barriers.

Overall, this is a big win for JuMP with minimal impact on other users. It's simple to implement, and there are no edge cases.

Reply:
I thing there will be a change in perf even for JuMP. Now there are two fields of abstract type.
We don't know what `backend` is, so have one hit, then you figure out it's `CachingOptimizer` so you make a call to it and then the `optimizer` field is abstract again so you get a second hit.

Reply:
There is no change in the JuMP behavior:
```Julia
julia> model = Model(Clp.Optimizer);

julia> backend(model)
MOIU.CachingOptimizer{MOI.AbstractOptimizer,MOIU.UniversalFallback{MOIU.Model{Float64}}}
in state EMPTY_OPTIMIZER
in mode AUTOMATIC
with model cache MOIU.UniversalFallback{MOIU.Model{Float64}}
  fallback for MOIU.Model{Float64}
with optimizer MOIB.LazyBridgeOptimizer{MOIU.CachingOptimizer{Clp.Optimizer,MOIU.UniversalFallback{MOIU.Model{Float64}}}}
  with 0 variable bridges
  with 0 constraint bridges
  with 0 objective bridges
  with inner model MOIU.CachingOptimizer{Clp.Optimizer,MOIU.UniversalFallback{MOIU.Model{Float64}}}
    in state ATTACHED_OPTIMIZER
    in mode AUTOMATIC
    with model cache MOIU.UniversalFallback{MOIU.Model{Float64}}
      fallback for MOIU.Model{Float64}
    with optimizer Clp.Optimizer

julia> model2 = Model(Clp.Optimizer; auto_bridge = true);

julia> backend(model2)
MOIU.CachingOptimizer{MOI.AbstractOptimizer,MOIU.UniversalFallback{MOIU.Model{Float64}}}
in state EMPTY_OPTIMIZER
in mode AUTOMATIC
with model cache MOIU.UniversalFallback{MOIU.Model{Float64}}
  fallback for MOIU.Model{Float64}
with optimizer Clp.Optimizer

julia> @variable(model2, x[1:2] in MOI.Nonnegatives(2))
2-element Array{VariableRef,1}:
 x[1]
 x[2]

julia> backend(model2)
MOIU.CachingOptimizer{MOI.AbstractOptimizer,MOIU.UniversalFallback{MOIU.Model{Float64}}}
in state EMPTY_OPTIMIZER
in mode AUTOMATIC
with model cache MOIU.UniversalFallback{MOIU.Model{Float64}}
  fallback for MOIU.Model{Float64}
with optimizer MOIB.LazyBridgeOptimizer{MOIU.CachingOptimizer{Clp.Optimizer,MOIU.UniversalFallback{MOIU.Model{Float64}}}}
  with 0 variable bridges
  with 0 constraint bridges
  with 0 objective bridges
  with inner model MOIU.CachingOptimizer{Clp.Optimizer,MOIU.UniversalFallback{MOIU.Model{Float64}}}
    in state ATTACHED_OPTIMIZER
    in mode AUTOMATIC
    with model cache MOIU.UniversalFallback{MOIU.Model{Float64}}
      fallback for MOIU.Model{Float64}
    with optimizer Clp.Optimizer
```

Reply:
If bridges are applied, we get back to the current JuMP behavior. If bridges are not applied, then we skip all the issues with bridges, but the backend type that JuMP sees is still the same.

Reply:
Can we have a call to discuss instead of this back-and-forth?

Reply:
Yes, I think we reached the point where messages with opposite timezone makes this process long and inefficient :D